### PR TITLE
Adds single-use event listener audit

### LINKED ISSUE #872: DBW: single-use event listeners instead of add/remove patterns
https://www.chromestatus.com/feature/5630331130478592

```
for all nodes
  gather event listeners

for each event listener, check if it is calling removeEventListener instead of itself (with the same params as it was created with it
```

Suggest `addEventListener('event', ... {once: true})` instead!

### PR DESCRIPTION
**Pull request by ebidel**
R: @paulirish @brendankenny @GoogleChrome/lighthouse 

Fixes #872

screen shot 2016-11-13 at 12 04 37 pm

### CODE REVIEW

**Comment by ebidel** [comment]
The smoke tests are failing because of https://github.com/GoogleChrome/lighthouse/pull/937/files#diff-8c4f14090828d17030843d8159c037bfR62, but this is expected.

**Comment by brendankenny** [comment]
> The smoke tests are failing because of https://github.com/GoogleChrome/lighthouse/pull/937/files#diff-8c4f14090828d17030843d8159c037bfR62, but this is expected.

change the expectations to `-1` in that case?

**Comment by brendankenny** [comment]
In the past we've waited for Chrome versions to hit stable before requiring them (like in #793), but since this can be sniffed it should be OK

**Comment by ebidel** [comment]
Why dont the smoketests use canary?

**Comment by brendankenny** [comment]
it could just be old travis caches. I'll try clearing and rerunning.

That is arguably a reason to hold off on this until m56 is stable, since people running stable won't be able to pass `npm run unit`

**Comment by ebidel** [comment]
Tests look 💯 now.
 
The `npm run unit` tests don't run chrome so this only affects (temporarily) the folks that check out lighthouse from src, don't have Canary installed, and that run the smoke tests. That number is probably very small. Chrome < 56 users will mostly be using the extension or CLI, which the audit protects against.

File: lighthouse-core/audits/dobetterweb/uses-single-use-event-listeners.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      return SingleUseEventsAudit.generateAuditResult({
+        rawValue: -1,
+        debugString: 'This audit requires Chrome 56+'
+      });
+    }
+
+    // Filter out event listeners that should be single-use.
+    const results = listeners.filter(loc => {
+      const removesOwnListener = loc.handler.description.match(
+          /\.removeEventListener\(/g);
```
Comment (brendankenny) [inline lighthouse-core/audits/dobetterweb/uses-single-use-event-listeners.js]: this should at least be checking the same event in the matching `addEventListener` statement. Even better would be matching the handler function itself added and removed. Would this gatherer be better using the call site method and comparing the two whenever there is a matching add and remove?

**Comment by ebidel** [comment]
> is there any advantage to using once in a world where you have to include the removeEventListener fallback anyways?

This is mostly a code ergonomics suggestion to raise awareness. FF 50 just landed it.

**Comment by ebidel** [comment]
@brendankenny can you take a look at this?

File: lighthouse-core/audits/dobetterweb/uses-single-use-event-listeners.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    // Note: EventListener.once landed in Chrome 56. Give up if we don't have it.
+    // https://codereview.chromium.org/2490303003
+    if (listeners.length && !('once' in listeners[0])) {
+      return SingleUseEventsAudit.generateAuditResult({
+        rawValue: -1,
+        debugString: 'This audit requires Chrome 56+'
+      });
+    }
+
+    // Filter out event listeners that should be single-use.
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/uses-single-use-event-listeners.js]: done

File: lighthouse-core/audits/dobetterweb/uses-single-use-event-listeners.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    const results = listeners.filter(loc => {
+      const regex = new RegExp(`\\.removeEventListener\\(\\s*['"]${loc.type}['"]\\s*,`, 'g');
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/uses-single-use-event-listeners.js]: Interesting data. Yea, now I'm thinking the regexp code check won't be that adequate. We could be more liberal and leave off the `loc.type` part? That may be overly noisy though. Any ideas?

**Comment by wardpeet** [comment]
@brendankenny your input it needed on this one.

@ebidel perhaps you can cleanup the chrome stable checks cause we are on 57 already 😄

**Comment by ebidel** [comment]
I'm ok with closing this. Not sure how useful it will be. Enough in the codebase has changed that we'll need to redo a bunch. Thoughts?

**Comment by paulirish** [comment]
im ok with closing too.

it's pretty leet but enters into a more proactive area that i think we're not ready to tackle just yet. (also browser support politics)

**Pull request closed without merging**